ads/google/a4a/line-delimited-response-handler.js

```diff
@@ -0,0 +1,116 @@
+ * limitations under the License.
+ */
+
+ import {dev} from '../../../src/log';
+ import {isObject} from '../../../src/types';
+
+ /**
+  * Creates an XHR streaming request with expected response format of repeated
+  * pairs of JSON object metadata then string HTML delimited by line returns.
+  *
+  * @param {!../../../src/service/xhr-impl.Xhr} xhr
+  * @param {string} url
+  * @param {!function(string, !Object<string, *>, boolean)} slotCallback called
+  *    with creative, metadata, and boolean indicating if completed.  Failure to
+  *    JSON parse metadata will cause empty object to be given.
+  * @param {?../../../src/service/xhr-impl.FetchInitDef=} opt_init
+  * @return {!Promise<!../../../src/service/xhr-impl.FetchResponse>} response,
+  *    note that accessing body will cause error as its consumed in streaming.
+  */
+ export function fetchLineDelimitedChunks(xhr, url, slotCallback, opt_init) {
+   return xhr.fetch(url, opt_init)
+     .then(response => {
+       let metadata;
+       let snippet = '';
+       const chunkCallback = (chunk, done) => {
+         const regex = /([^\n]*)(\n)?/g;
+         let match;
+         while ((match = regex.exec(chunk))) {
+           snippet += match[1];
+           if (match[2]) {
+             if (metadata) {
+               slotCallback(unescapeHtml_(snippet), metadata,
+                  done && regex.lastIndex === chunk.length);
+               metadata = undefined;
+             } else {
+               metadata = safeJsonParse_(snippet);
+             }
+             snippet = '';
+           }
+           if (regex.lastIndex === chunk.length) {
+             break;
+           }
+         }
+       };
+       if (!response.body || !xhr.win.TextDecoder) {
+         // TODO(keithwrightbos) - TextDecoder polyfill?
+         response.text().then(content => chunkCallback(content, true));
+       } else {
+         handleFetchResponseStream_(
+           response.body.getReader(), chunkCallback, new TextDecoder('utf-8'));
+       }
+       return {status: response.status, headers: response.headers};
+     });
+ }
+
+ /**
+  * @param {!ReadableStreamReader} reader The reader of the fetch's
+  *  response stream.
+  * @param {!function(string, boolean)} chunkHandler
+  * @param {!TextDecoder} textDecoder
+  * @private
+  */
+ function handleFetchResponseStream_(reader, chunkHandler, textDecoder) {
+   reader.read().then(function chunk(result) {
+     if (result.value) {
+       chunkHandler(
+         textDecoder.decode(
+           /** @type {!ArrayBuffer} */(result.value), {'stream': true}),
+         result.done);
+     }
+     if (!result.done) {
+       // More chunks to read.
+       reader.read().then(chunk);
+     }
+   });
+ }
+
+ /**
+  * Unescapes characters that are escaped in line-delimited JSON-HTML.
+  * @param {string} html An html snippet.
+  * @return {string}
+  * @private
+  */
+ function unescapeHtml_(html) {
```

Comment:
Maybe name can be unescapeLineDelimitedHtml_ instead, would make it clearer for people reading that they aren't seeing an AMP specific version of UriDecode or something.